### ISSUE TITLE
False-positive unused variable warning in the presence of `unimplemented!()`

### ISSUE THREAD

**Issue by nzentzis** [rust-lang/rust#88317]
Given the following code: [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3d57bbf18fb502de8991d1f1b5affe09)

```rust
pub enum View {
    A, B
}

pub struct Foo;

impl Foo {
    pub fn do_stuff(&self, view: View) {
        let _: f32 = unimplemented!();
        match view {
            _ => {},
        }
    }
}
```

The current output warns about an unused parameter:

```
warning: unused variable: `view`
 --> src/lib.rs:8:28
  |
8 |     pub fn do_stuff(&self, view: View) {
  |                            ^^^^ help: if this is intentional, prefix it with an underscore: `_view`
  |
  = note: `#[warn(unused_variables)]` on by default
```

However, the `view` parameter is used in the match block, even though it'll never be executed as things stand. If I comment out the `unimplemented!()`, the warning goes away.

**Comment by LiamTheProgrammer** [comment]
I thought this was intended behavior? The variable really is never used when the codes runs...

**Comment by ehuss** [comment]
Thanks for the report!  This is intended behavior, as the variables really aren't used since the "used" parts are in an unreachable block.  Closing as a duplicate of #78136.
